### ISSUE TITLE
Signing and verifying a message with eToken/Improving error handling

### ISSUE THREAD

**Issue by Hyperiongit** [PeculiarVentures/graphene#68]
So, I have this following piece of code that I came up with, mostly using your provided examples, and I have two questions.

```
var graphene = require("graphene-pk11");
var Module = graphene.Module;

var mod = Module.load("H://Windows/System32/etpkcs11.dll", "eToken");

mod.initialize();

var session = mod.getSlots(0).open();
session.login("123");

// Get all objects from the eToken
var arr = session.find();
if(arr.length > 0){
   for(var i=0; i < arr.length;i++){        
      if(arr.items(i).class == graphene.ObjectClass.PRIVATE_KEY){  
        //Get a PRIVATE_KEY object for signing and initialize a sign object     
        var pkey = arr.items(i).toType(graphene.ObjectClass.PRIVATE_KEY);      
        var sign = session.createSign("SHA1_RSA_PKCS",pkey); 
        sign.update("simple text 1");    
        sign.update("simple text 2");

        //sign the text and print it     
        var signature = sign.final();
        console.log("Signature RSA-SHA1:", signature.toString("hex"));

        //get the publicExponent and modulus from the PRIVATE_KEY object so we
        //can create a PUBLIC_KEY object for verifying the signature      
        var okey = pkey.getAttribute({modulus: null, publicExponent: null});

        //PUBLIC_KEY template basis is taken from 
        //https://www.cryptsoft.com/pkcs11doc/v230/group__SEC__11__1__2__RSA__PUBLIC__KEY__OBJECTS.html
        var rpkey = session.create({class: graphene.ObjectClass.PUBLIC_KEY, //CKA_CLASS
                                    keyType: graphene.KeyType.RSA, //CKA_KEY_TYPE
                                    token: false, //CKA_TOKEN
                                    label: "temp", //CKA_LABEL
                                    wrap: true, //CKA_WRAP
                                    encrypt: true, //CKA_ENCRYPT
                                    modulus: okey.modulus, //CKA_MODULUS
                                    publicExponent: okey.publicExponent}); //CKA_PUBLIC_EXPONENT
        
        //create verification object and initialize it with our PUBLIC_KEY
        var verify = session.createVerify("SHA1_RSA_PKCS", rpkey);
        verify.update("simple text 1");
        verify.update("simple text 2");

        //finalize verification and print result     
        var verify_result = verify.final(signature);
        console.log("Signature RSA-SHA1 verify:", verify_result);
        
        //destroying our PUBLIC_KEY object      
        rpkey.destroy();
        //do not continue for other private keys on the token
        break;
      }
   }        
}
session.logout();
mod.finalize();
```

Questions:
1) Function session.find() doesn't find any PUBLIC_KEY objects, therefore in the code above, the PUBLIC_KEY object is created outside of eToken from several attributes of PRIVATE_KEY, using function session.create(). 
I'm not sure WHERE it is created though, and if it's purely a temporary object on the heap that would be cleaned up by the OS after the process exit if it's not destroyed by destroy() function. 
The question is: does this code implement the correct way of signing and verifying a message with a token, using graphene?
2) This is the output for correct message signature:
_Signature RSA-SHA1: 9b3bdd9eeecf4fe1e8dff162b4bd4b090e1948ad7362a2e1de4479e69f11
fc9500a1dbcfb2bd1208ff2d987afa6a8226f5a3e54e3ab4fcfb6d97b979fba8857b33dfde47485a
417659d8a8f36c6fdf75f347b67b0d23e6020749b2ddc63d547bfae027976584fb68ad630fdfc11f
81141c8704c8525504e0f264a65a921151d40488ef59f9f0739c1c8a3ac606d57ce4672a758d77ae
228b092cabc9fca2423065304c92c530dd189eb640212017e52d7ef7f0a55b297c473492bf65821c
5357f98101470a9371e5aaa2bd069356a4490033dc059c68489804b8c5ffee82ed4a9bc6245dd98c
4ab95ccf62b81f9ad31d53bfbfcb3026a2a112f63a3f65b44b28
Signature RSA-SHA1 verify: true_

However, if we change the text from "simple text 2" to, say, "simple text 3"  for verification, we get this:

H:\Users\R\node_modules\graphene-pk11\build\crypto\verify.js:35
        var res = this.lib.C_VerifyFinal(this.session.handle, signature);
                           ^
Error: CKR_SIGNATURE_INVALID:192
    at Error (native) PKCS11::C_VerifyFinal:837
    at Error (native)
    at Verify.final (H:\Users\R\node_modules\graphene-pk11\build\crypto\verify.j
s:35:28)
    at Object.<anonymous> (H:\Users\R\nsfg.js:48:36)
    at Module._compile (module.js:570:32)
    at Object.Module._extensions..js (module.js:579:10)
    at Module.load (module.js:487:32)
    at tryModuleLoad (module.js:446:12)
    at Function.Module._load (module.js:438:3)
    at Module.runMain (module.js:604:10)
    at run (bootstrap_node.js:394:7)

So basically, verification function returns true if the signature is correct and throws an exception if it's not. Which seems to be a weird way of implementing it (Shouldn't it just return false? Does user need this exception?). The question is, can this be improved?

Thanks in advance.

**Comment by microshine** [comment]
1. Creating keys
The right way to create key is https://github.com/PeculiarVentures/graphene#generating-keys
`token: false, //CKA_TOKEN` - key (object) will be created in mamory. If you set `true` it'll be stored on token and you can get it on next session

2. You are right. `CKR_SIGNATURE_INVALID` must return `false`. I'll fix it later

**Comment by Hyperiongit** [comment]
Thank you for the answer.

I'm not generating a key though, I'm trying to use existing ones, from the token. The find function gives me only PRIVATE_KEY objects on the token, and I don't see how I can get a PUBLIC_KEY object(for verifying) from it's PRIVATE_KEY counterpart (that I get from session.find) using generateKey. Any example on how to get a PUBLIC_KEY from an existing PRIVATE_KEY object in a "correct" way?

**Comment by microshine** [comment]
I think your example code is right, but I can check it later

RSA private keys MAY return `CKA_MODULUS`, `CKA_PUBLIC_EXPONENT`
https://www.cryptsoft.com/pkcs11doc/v230/group__SEC__11__1__4__PKCS____1__RSA__KEY__PAIR__GENERATION.html#CKM_RSA_PKCS_KEY_PAIR_GEN"
